app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/UpscanController.scala

```diff
@@ -0,0 +1,148 @@
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class UpscanController @Inject()(
+  subscriptionService: SubscriptionService,
+  upscanService: UpscanService,
+  upscanConnector: UpscanConnector,
+  configuration: Configuration,
+  sessionStore: SessionStore,
+  errorHandler: ErrorHandler,
+  cc: MessagesControllerComponents,
+  val authenticatedAction: AuthenticatedAction,
+  val sessionDataAction: SessionDataAction,
+  upscanPage: html.upscan.upscan,
+  upscanLimitPage: html.upscan.upscan_limit,
+  upscanSuccessPage: html.upscan.upscan_success
+)(implicit viewConfig: ViewConfig, ec: ExecutionContext)
+    extends FrontendController(cc)
+    with WithAuthAndSessionDataAction
+    with Logging
+    with SessionUpdates {
+
+  private def withSubscribedUser(request: RequestWithSessionData[_])(
+    f: (SessionData, Subscribed) => Future[Result]
+  ): Future[Result] =
+    request.sessionData.flatMap(s => s.journeyStatus.map(s -> _)) match {
+      case Some((s: SessionData, r: Subscribed)) =>
+        f(s, r)
+      case _ =>
+        Future.successful(
+          SeeOther(uk.gov.hmrc.cgtpropertydisposalsfrontend.controllers.routes.StartController.start().url)
+        )
+    }
+
+  private def getUpscanInitiateConfig[A: Configs](key: String): A =
+    configuration.underlying
+      .get[A](s"microservice.services.upscan-initiate.$key")
+      .value
+
+  private val maxFileSize: Int = getUpscanInitiateConfig[Int]("max-uploads")
+
+  @SuppressWarnings(Array("org.wartremover.warts.Var", "org.wartremover.warts.Any"))
+  def upload(): Action[MultipartFormData[Files.TemporaryFile]] = Action(parse.multipartFormData(maxFileSize)).async {
+    implicit request: MessagesRequest[MultipartFormData[Files.TemporaryFile]] =>
+      val multipart = request.body
+      multipart.dataParts.get("href").flatMap(_.headOption).fold(Future.successful(BadRequest("missing href"))) {
+        href =>
+          val data: Map[String, Seq[String]] = multipart.dataParts.map {
+            case (k, v) => (k.stripPrefix("upscan."), v)
+          }
+          val withSource =
+            multipart.files.map(file => file.copy(ref = FileIO.fromPath(file.ref.path): Source[ByteString, Any]))
+          val prepared = multipart.copy(files = withSource, dataParts = data)
+          upscanConnector.upload(href, prepared).value.map {
+            case Left(error) =>
+              logger.warn(s"Could not upload file to S3 due to $error")
+              errorHandler.errorResult(None)
+            case Right(_) => Ok
+          }
+      }
+  }
+
+  def errorRedirect(cgtReference: String): Action[AnyContent] = Action {
+    implicit request: MessagesRequest[AnyContent] =>
+      errorHandler.errorResult(None)
+  }
+
+  def successRedirect(cgtReference: String): Action[AnyContent] = Action {
+    implicit request: MessagesRequest[AnyContent] =>
+      Ok(upscanSuccessPage())
```

Comment:
we'd never see this page rendered anywhere right since this is triggered via a callback from upscan? what responses do they expect back from the callbacks or can we reply back with anything?

Reply:
it depends - the UX team need to design this user flow.

Reply:
the callback is asynchronous so these pages would never render - I imagine when we come to do the file upload in our journey we'd make the callback update some flag in mongo somewhere which will indicate that the file upload was successful. I'm imagining we'd have something like:
```scala
case class AttachedFile(id: ID, upscanSuccess: Option[Boolean])
case class DraftReturn(..., attachedFiles: List[AttachedFile])
```
with draft returns stored somewhere.

Reply:
what you have now isn't broken and we can't really do anything properly with the callbacks atm since we don't have a draft return store or anything. For now though it'd be useful to see a log message if it was a success or failure with the body of the request body logged just so we can see what's happening when we test it.

Reply:
oh wait a sec, this isn't the callback. Ignore me :D

Reply:
:smiley: - the upscan service is badly designed tho - browser redirects :man_facepalming:

Reply:
I have reverted back to v1 initiate. This redirect feature of upscan is no good and we don't really need it. V1 allows us to handle success/error statuses from within the CGT service. Also with v2, there would be no way to block users from uploading files greater than 5 MB unless client side JS is added. For users who don't have JS enabled in their browser this could potentially lead to DMS failures.